### PR TITLE
Added version diff api endpoint

### LINKED ISSUE #3659: Creation version diff API
The goal of this ticket is to create an API that allows to compare desired state versions.

Based on the initial exploration this API should do the following:

- Accept two versions. In the PoC called r1 and r2. The diff is the "editscript" that transforms r1 into r2
- The backend calculates per resources and per attribute what the diff is. When the attribute does not exist in a version, it is set to null.
   - See https://github.com/inmanta/web-console/issues/2244 for details on the calculation
   - The API generates two values: the actual value and the value that can be used for compare (string)
   - This script does a number of conversions (see diff.py in PoC ticket zip file)
      - A list is converted in text with one item per line
      - json is stringified
      - requires are treated as a list but versions are stripped so that they can be compared
      - numbers and bool are converted to text
      - see also compare value section (the guessed type is actually not used any more)
- The frontend generates the diff based on the compare values of the attribute
- There can be many resources and lots of large values (duplication due to value and compare value) so the endpoint needs paging
- the resource has a status field with value 'added', 'updated' or 'deleted'
- no paging

**compare value**

```
    def guess_type(self) -> str:
        """Guess the type of the data and based on that guess generate a value that can be used for comparison"""
        if self._type is not None:
            return self._type

        if isinstance(self._value, list) and len(self._value) > 0 and not isinstance(self._value[0], (dict, list)):
            # sort list that are not a json entry
            self._type = "list"
            self._compare_value = "\n".join([str(x) for x in sorted(self._value)])

        elif isinstance(self._value, (dict, list)):  # json
            self._type = "json"
            self._compare_value = json.dumps(self._value, indent=4, sort_keys=True)
        else:
            self._type = "simple"
            self._compare_value = str(self.value)

        return self._type
```

## API Example

```json
[
    {
        "resource_id": "kubernetes::resources::ConfigMap[dc-1,identifier=/cluster/dc-1/namespace/control-plane-1/config-map/open5gs-smf-config-map]",
        "status": "modified", # added or deleted 
        "attributes": {
            "data": {
                "r1": {
                    "smf.yaml": "logger:\n  file: /var/log/open5gs/smf.log\nnrf:\n  sbi:\n  - addr: 10.1.10.1\n    port: 7777\nparameter:\n  no_ipv6: true\n  use_openair: true\nsmf:\n  dns:\n  - 8.8.8.8\n  - 8.8.4.4\n  gtpc:\n  - addr: 10.1.10.5\n  gtpu:\n  - addr: 10.1.10.5\n  mtu: 1300\n  pfcp:\n  - addr: 10.1.10.5\n  sbi:\n  - addr: 10.1.10.5\n    port: 7777\n  subnet: []\nupf:\n  pfcp:\n  - addr: 10.1.12.2\n    nr_cell_id:\n    - '000000012'\n  - addr: 10.1.11.2\n    nr_cell_id:\n    - '000000011'\n",
                    "diameter-smf.conf": ""
                },
                "r2": {
                    "smf.yaml": "logger:\n  file: /var/log/open5gs/smf.log\nnrf:\n  sbi:\n  - addr: 10.1.10.1\n    port: 7777\nparameter:\n  no_ipv6: true\n  use_openair: true\nsmf:\n  dns:\n  - 8.8.8.8\n  - 8.8.4.4\n  gtpc:\n  - addr: 10.1.10.5\n  gtpu:\n  - addr: 10.1.10.5\n  mtu: 1300\n  pfcp:\n  - addr: 10.1.10.5\n  sbi:\n  - addr: 10.1.10.5\n    port: 7777\n  subnet:\n  - addr: 10.2.0.1/24\n    dnn: internet1\n  - addr: 10.2.1.1/24\n    dnn: internet2\nupf:\n  pfcp:\n  - addr: 10.1.12.2\n    nr_cell_id:\n    - '000000012'\n  - addr: 10.1.11.2\n    nr_cell_id:\n    - '000000011'\n",
                    "diameter-smf.conf": ""
                },
                "r1_compare": "{\n    \"diameter-smf.conf\": \"\",\n    \"smf.yaml\": \"logger:\\n  file: /var/log/open5gs/smf.log\\nnrf:\\n  sbi:\\n  - addr: 10.1.10.1\\n    port: 7777\\nparameter:\\n  no_ipv6: true\\n  use_openair: true\\nsmf:\\n  dns:\\n  - 8.8.8.8\\n  - 8.8.4.4\\n  gtpc:\\n  - addr: 10.1.10.5\\n  gtpu:\\n  - addr: 10.1.10.5\\n  mtu: 1300\\n  pfcp:\\n  - addr: 10.1.10.5\\n  sbi:\\n  - addr: 10.1.10.5\\n    port: 7777\\n  subnet:\\n  - addr: 10.2.0.1/24\\n    dnn: internet1\\n  - addr: 10.2.1.1/24\\n    dnn: internet2\\nupf:\\n  pfcp:\\n  - addr: 10.1.12.2\\n    nr_cell_id:\\n    - '000000012'\\n  - addr: 10.1.11.2\\n    nr_cell_id:\\n    - '000000011'\\n\"\n}",
                "r2_compare": "{\n    \"diameter-smf.conf\": \"\",\n    \"smf.yaml\": \"logger:\\n  file: /var/log/open5gs/smf.log\\nnrf:\\n  sbi:\\n  - addr: 10.1.10.1\\n    port: 7777\\nparameter:\\n  no_ipv6: true\\n  use_openair: true\\nsmf:\\n  dns:\\n  - 8.8.8.8\\n  - 8.8.4.4\\n  gtpc:\\n  - addr: 10.1.10.5\\n  gtpu:\\n  - addr: 10.1.10.5\\n  mtu: 1300\\n  pfcp:\\n  - addr: 10.1.10.5\\n  sbi:\\n  - addr: 10.1.10.5\\n    port: 7777\\n  subnet:\\n  - addr: 10.2.0.1/24\\n    dnn: internet1\\n  - addr: 10.2.1.1/24\\n    dnn: internet2\\nupf:\\n  pfcp:\\n  - addr: 10.1.12.2\\n    nr_cell_id:\\n    - '000000012'\\n  - addr: 10.1.11.2\\n    nr_cell_id:\\n    - '000000011'\\n\"\n}",
            }
        }
    },
]
```

## Questions:
- Are added and removed resource entirely implicit?
- Do we persist compare values? -> only if required.

### PR DESCRIPTION
**Pull request by andraskvr**
# Description

closes #3659 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/server/diff.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        diff = {}
+        # Generate the diff for each attribute
+        for added in our_attributes - other_attributes:
+            diff[added] = self._attributes[added].added()
+
+        for removed in other_attributes - our_attributes:
+            diff[removed] = other._attributes[removed].removed()
+
+        for name in other_attributes.intersection(our_attributes):
+            attr_diff = self._attributes[name].compare(other._attributes[name])
```
Comment (wouterdb) [inline src/inmanta/server/diff.py:110]: I would prefer the more symmetrical solution of implementing __sub__ and __rsub__ on `Attribute` but that is probably just me. 

```
for k in other_attribute + our_attributes:
   attr_diff = self._attributes.get(name) - other._attributes.get(name)
   if attr_diff is not None:
                diff[name] = attr_diff
```

so this is just a comment, no need to change it.

**Review by bartv — CHANGES_REQUESTED** [review]
It is probably more useful for Sander or Arnaud to review this as well, instead of me. My main comment is a classic: methods and classes that are not dead obvious should have a comment to provide content.

Maybe even more important is adding documentation to the test case. It is far from trivial. In 2 years all context on what is tested has faded again.

**Review by wouterdb — APPROVED** [review]

**Review by bartv — APPROVED** [review]

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/server/diff.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+        return self._compare_value
+
+    def _generate_compare_value(self) -> None:
+        """Generate a value that can be used for comparison"""
+        if self._compare_value is not None:
+            return
+
+        if isinstance(self._value, list) and len(self._value) > 0 and not isinstance(self._value[0], (dict, list)):
+            self._compare_value = "\n".join([str(x) for x in self._value])
```
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/diff.py:55]: Hmm. `sort_keys` has only effect on dicts. So that's not relevant.

[the remark this answers is not in the document]
Comment (andraskvr) [reply src/inmanta/server/diff.py]: I see your point and made the changes to use the json serializer for the lists as well.
Is that ok with you @bartv ?

[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/server/diff.py]: that is indeed a good idea

**Review by arnaudsjs — APPROVED** [review]

**Pull request closed without merging**